gk/main.c

```diff
@@ -679,207 +679,379 @@ static int
+				print_flow_err_msg(&packet.flow,
+					"gk: failed to get the fib entry");
+				drop_packet(pkt);
+			}
+			continue;
+		}
 
-		process_pkts_ipv6_acl(&gk_conf->net->front, lcore, &acl);
+		switch (fib->action) {
+		case GK_FWD_FRONT_NET: {
+			/*
+			 * The entry instructs to forward
+			 * its packets to the front interface,
+			 * forward accordingly.
+			 *
+			 * BP bypasses from the back to the front interface
+			 * are expected to bypass the outgoing traffic
+			 * from the AS to its peers.
+			 *
+			 * Notice that one needs to update
+			 * the Ethernet header.
+			 *
+			 * TODO Add sequential lock to deal with the
+			 * concurrency between GK and LLS on the cached
+			 * Ethernet header.
+			 */
+			struct ether_cache *eth_cache =
+				&fib->u.next_fib->u.nexthop.eth_cache;
+			struct ether_hdr *eth_hdr =
+				rte_pktmbuf_mtod(
+				pkt, struct ether_hdr *);
+
+			if (eth_cache->stale) {
+				drop_packet(pkt);
+				continue;
+			}
+
+			rte_memcpy(eth_hdr,
+				&eth_cache->eth_hdr, sizeof(*eth_hdr));
+			tx_bufs[num_tx++] = pkt;
+			continue;
+		}
+
+		default:
+			/* All other actions should log a warning. */
+			RTE_LOG(WARNING, GATEKEEPER,
+				"gk: the fib entry has an unexpected action %u at %s!\n",
+				fib->action, __func__);
+			drop_packet(pkt);
+			continue;
+		}
+	}
+
+	/* Send burst of TX packets, to second port of pair. */
+	num_tx_succ = rte_eth_tx_burst(port_front, tx_queue_front,
+		tx_bufs, num_tx);
+
+	/* XXX Do something better here! For now, free any unsent packets. */
+	if (unlikely(num_tx_succ < num_tx)) {
+		for (i = num_tx_succ; i < num_tx; i++)
+			rte_pktmbuf_free(tx_bufs[i]);
+	}
+}
+
+static void
+process_cmds_from_mailbox(
+	struct gk_instance *instance, struct gk_config *gk_conf)
+{
+	int i;
+	int num_cmd;
+	struct gk_cmd_entry *gk_cmds[GK_CMD_BURST_SIZE];
+
+	/* Load a set of commands from its mailbox ring. */
+        num_cmd = mb_dequeue_burst(&instance->mb,
+               	(void **)gk_cmds, GK_CMD_BURST_SIZE);
+
+        for (i = 0; i < num_cmd; i++) {
+		process_gk_cmd(gk_cmds[i], instance, &gk_conf->lpm_tbl);
+		mb_free_entry(&instance->mb, gk_cmds[i]);
+        }
+}
+
+static int
+gk_proc(void *arg)
+{
+	/* TODO Implement the basic algorithm of a GK block. */
+
+	unsigned int lcore = rte_lcore_id();
+	struct gk_config *gk_conf = (struct gk_config *)arg;
+	unsigned int block_idx = get_block_idx(gk_conf, lcore);
+	struct gk_instance *instance = &gk_conf->instances[block_idx];
+
+	uint8_t port_front = get_net_conf()->front.id;
+	uint8_t port_back = get_net_conf()->back.id;
+	uint16_t rx_queue_front = instance->rx_queue_front;
+	uint16_t tx_queue_front = instance->tx_queue_front;
+	uint16_t rx_queue_back = instance->rx_queue_back;
+	uint16_t tx_queue_back = instance->tx_queue_back;
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"gk: the GK block is running at lcore = %u\n", lcore);
+
+	gk_conf_hold(gk_conf);
+
+	while (likely(!exiting)) {
+		IPV6_ACL_SEARCH_DEF(acl_front);
+
+		IPV6_ACL_SEARCH_DEF(acl_back);
```

Comment:
What's the motivation of bringing `acl_front` and `acl_back` into `gk_proc()` instead of leaving them in `process_pkts_front()` and `process_pkts_back()`, respectively?